### PR TITLE
Vmware offline migration

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

Enable storage migration for stopped VMs on Vmware and for detached volumes on primary storages in Vmware clusters.

see https://cwiki.apache.org/confluence/display/CLOUDSTACK/VMware+offline+migrations

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
travis red, but the travis job says all green ???

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware65

**Comment by rhtyd** [comment]
all environments have failed deployment, see sblab for errors @DaanHoogland

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@mike-tutkowski @rafaelweingartner @rhtyd as you guys are on a dull conference anyway, could you review this one ;)

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-55u3

**Comment by dhlaluku** [comment]
@blueorangutan test matrix

**Comment by dhlaluku** [comment]
@nvazquez @rhtyd @rafaelweingartner @mike-tutkowski @borisstoyanov please review

**Comment by mike-tutkowski** [comment]
@DaanHoogland Somehow I didn't notice the request for review here, Daan. I should be able to get to this soon.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java [hunk 1]
```diff
@@ -192,4 +187,25 @@ public void execute() {
             throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, e.getMessage());
         }
     }
+
+    @Override
+    public String getSyncObjType() {
```
Comment (rafaelweingartner) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: Why did you need to implemente these `getSyncObjType` and `getSyncObjId`? Could you explain this a little bit further?

File: api/src/main/java/org/apache/cloudstack/api/command/user/volume/MigrateVolumeCmd.java [hunk 2]
```diff
@@ -120,4 +120,19 @@ public void execute() {
...
+    public String getSyncObjType() {
+        if (getSyncObjId() != null) {
+            return BaseAsyncCmd.migrationSyncObject;
+        }
+        return null;
+    }
+
+    @Override
+    public Long getSyncObjId() {
+        if (getStoragePoolId() != null) {
```
Comment (rafaelweingartner) [inline api/src/main/java/org/apache/cloudstack/api/command/user/volume/MigrateVolumeCmd.java]: If `getStoragePoolId()` != null, then return `getStoragePoolId()`, otherwise return null....

You can simple call `return getStoragePoolId()`. The result is the same.

File: core/src/main/java/com/cloud/agent/api/MigrateVmToPoolCommand.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    protected MigrateVmToPoolCommand() {
+    }
+
+    /**
+     *
+     * @param vmName the name of the VM to migrate
+     * @param volumes used to supply feedback on vmware generated names
+     * @param destinationPool the primary storage pool to migrate the VM to
+     * @param executeInSequence
```
Comment (rafaelweingartner) [inline core/src/main/java/com/cloud/agent/api/MigrateVmToPoolCommand.java]: if you do not have anything further to talk about a parameter, I think that you do not need to list it here.

File: core/src/main/java/com/cloud/agent/api/UnregisterVMCommand.java [hunk 4]
```diff
@@ -22,14 +22,19 @@
 public class UnregisterVMCommand extends Command {
     String vmName;
     boolean cleanupVmFiles = false;
+    boolean executeInSequence = false;
```
Comment (rafaelweingartner) [inline core/src/main/java/com/cloud/agent/api/UnregisterVMCommand.java]: The default is `false`. No need to declare it.

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java [hunk 5]
```diff
@@ -25,11 +25,28 @@
 import com.cloud.agent.api.to.VirtualMachineTO;
 import com.cloud.host.Host;
 
+/**
+ * interface to query how to move data around and to commision the moving
```
Comment (rafaelweingartner) [inline engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: I am loving this new approach that you are adopting to document the code you introduce in ACS 👍

File: engine/components-api/src/main/java/com/cloud/storage/StorageManager.java [hunk 6]
```diff
@@ -106,7 +106,14 @@
      * @param poolId
      * @return comma separated list of tags
      */
-    public String getStoragePoolTags(long poolId);
+    String getStoragePoolTags(long poolId);
+
+    /**
+     * Returns a list of Strings with tags for the specified storage pool
+     * @param poolId
```
Comment (rafaelweingartner) [inline engine/components-api/src/main/java/com/cloud/storage/StorageManager.java]: If you do not have anything to say about a parameter, you do not need to declare it here.

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java [hunk 7]
```diff
@@ -25,11 +25,28 @@
...
+     * @param srcData object to move
+     * @param destData location to move it to
+     * @return the expertise level with which this instance knows how to handle the move
+     */
     StrategyPriority canHandle(DataObject srcData, DataObject destData);
 
     StrategyPriority canHandle(Map<VolumeInfo, DataStore> volumeMap, Host srcHost, Host destHost);
 
+    /**
+     * copy the source volume to its destination (on a host if not null)
```
Comment (rafaelweingartner) [inline engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: what happens if the host is null?

Is it possible to be null when we get to this part of the code?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 8]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
     private void orchestrateStorageMigration(final String vmUuid, final StoragePool destPool) {
         final VMInstanceVO vm = _vmDao.findByUuid(vmUuid);
 
+        preStorageMigrationStateCheck(destPool, vm);
+
+        try {
+            if(s_logger.isDebugEnabled()) {
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: do you really need to check if debug log level is enabled? You can simply log it, and let the logging framework to manage it.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 9]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+                                vm.getInstanceName()));
+            }
+
+            migrateThroughHypervisorOrStorage(destPool, vm);
+
+        } catch (ConcurrentOperationException
+                | InsufficientCapacityException // possibly InsufficientVirtualNetworkCapacityException or InsufficientAddressCapacityException
+                | StorageUnavailableException e) {
+            s_logger.debug("Failed to migration: " + e.toString());
+            throw new CloudRuntimeException("Failed to migration: " + e.toString());
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: This will break the stack trace. So, what about `throw new CloudRuntimeException(Sring.format ("Failed to migrate VM: %s ", vm.getUuid()), e);` instead?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 10]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+        } finally {
+            try {
+                stateTransitTo(vm, Event.AgentReportStopped, null);
+            } catch (final NoTransitionException e) {
+                s_logger.debug("Failed to change vm state: " + e.toString());
+                throw new CloudRuntimeException("Failed to change vm state: " + e.toString());
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: The same thing here

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 11]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+            }
+        }
+    }
+
+    private Answer[] attemptHypervisorMigration(StoragePool destPool, VMInstanceVO vm) {
+        boolean migrationResult = false;
+        final HypervisorGuru hvGuru = _hvGuruMgr.getGuru(vm.getHypervisorType());
+        // OfflineVmwareMigration: in case of vmware call vcenter to do it for us.
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: instead of commenting here, what about a proper Java doc?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 12]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+        // OfflineVmwareMigration: should we check the proximity of source and destination
+        // OfflineVmwareMigration: if we are in the same cluster/datacentre/pool or whatever?
+        // OfflineVmwareMigration: we are checking on success to optionally delete an old vm if we are not
+        List<Command> commandsToSend = hvGuru.finalizeMigrate(vm, destPool);
+
+        Long hostId = vm.getHostId();
+        // OfflineVmwareMigration: probaby this is null when vm is stopped
+        if(hostId == null) {
+            hostId = vm.getLastHostId();
+            if (s_logger.isDebugEnabled()) {
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: do we really need this?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 13]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+                s_logger.debug(String.format("host id is null, using last host id %d", hostId) );
+            }
+        }
+
+        if(CollectionUtils.isNotEmpty(commandsToSend)) {
+            Commands commandsContainer = new Commands(commandsToSend.get(0));
+            try {
+                // OfflineVmwareMigration: change to the call back variety?
+                // OfflineVmwareMigration: getting a Long seq to be filled with _agentMgr.send(hostId, commandsContainer, this)
+                Answer[] answers = _agentMgr.send(hostId, commandsContainer);
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: `return _agentMgr.send(hostId, commandsContainer);`

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 14]
```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+        // OfflineVmwareMigration: probaby this is null when vm is stopped
+        if(hostId == null) {
+            hostId = vm.getLastHostId();
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("host id is null, using last host id %d", hostId) );
+            }
+        }
+
+        if(CollectionUtils.isNotEmpty(commandsToSend)) {
+            Commands commandsContainer = new Commands(commandsToSend.get(0));
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: are you only sending the first command?

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java (hunk 1 above)
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: i can, but it will be not a little bit ;)
tl;dr: limitting the number of jobs send to vmware.
There is a mechanism to limit jobs. It checks in the sync phase of async jobs to see if the queue for those jobs exeeds a certain length. If so it return to the user to try again later. It can filter on type and entity, if it can get them. These methods are part of those mechs.

File: api/src/main/java/org/apache/cloudstack/api/command/user/volume/MigrateVolumeCmd.java (hunk 2 above)
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/user/volume/MigrateVolumeCmd.java]: the consideration here (and the intermediate version) that other objects, like the host, could be the sync object as well so this could be extended to return that under conditions.

File: core/src/main/java/com/cloud/agent/api/MigrateVmToPoolCommand.java (hunk 3 above)
Comment (DaanHoogland) [reply core/src/main/java/com/cloud/agent/api/MigrateVmToPoolCommand.java]: doesn't that make it part of the javadoc rendering? it seems weird to include only part of the parameter list in the javadoc.

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java (hunk 5 above)
Comment (DaanHoogland) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: yes, i was feeling preppy ;) and in this case, I think it does say something more than the name implies.

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java (hunk 7 above)
Comment (DaanHoogland) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: maybe not in this PR but it is part of the bigger framework for data motion.

File: engine/components-api/src/main/java/com/cloud/storage/StorageManager.java (hunk 6 above)
Comment (DaanHoogland) [reply engine/components-api/src/main/java/com/cloud/storage/StorageManager.java]: I think when a name is self explanatory don't explain. You'd run the risk of confusing. However when making javadoc, do include all parameters, also to not confuse things. milage and religious views may vary

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 8 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: yes, you do if you are passing dynamic parameters

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 9 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: good point however the throwing is old code that was refactored for the multiple catch only. this is how the exception is thrown in the current system.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 12 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: String.format() should not be called for a string we are not logging anyway, so yes

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 14 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: good point, yes we are actually but the framework dictates. we might need to change this on a refactor though.

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test matrix

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan centos7 vmware-65

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 8 above)
Comment (mike-tutkowski) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Does the logging framework not allow for lambda expressions to be passed in? That would be the ideal way to handle dynamic parameters without have to put an if statement in.

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java (hunk 7 above)
Comment (dhlaluku) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: @rafaelweingartner StorageMotionStrategy that supports host to be null is selected and used for migrations, e.g. VmwareStorageStrategyMotion used by this PR/feature.

A non-blank host is only used for copying objects between primary storage pools.

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 8 above)
Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: @rafaelweingartner @mike-tutkowski removed the log level check and formatted the dynamic string prior to sending it as Log message 
String msg = String.format("Offline migration of %s vm %s with volumes", vm.getHypervisorType().toString(), vm.getInstanceName());
             s_logger.debug(msg);

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 9 above)
Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Refactored to use String.format(....)

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 8 above)
Comment (rafaelweingartner) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: @mike-tutkowski the current log4j we use does not allow that. However, log4j-2 does. I am trying to make time to finish the upgrade.

@dhlaluku even though I think that using the check `s_logger.isDebugEnabled()` is foolish in our context, you do not need to remove just because of me. I understand @DaanHoogland perspective.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 15]
```diff
@@ -1988,92 +1977,249 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+            }
+
+            migrateThroughHypervisorOrStorage(destPool, vm);
+
+        } catch (ConcurrentOperationException
+                | InsufficientCapacityException // possibly InsufficientVirtualNetworkCapacityException or InsufficientAddressCapacityException
+                | StorageUnavailableException e) {
+            String msg = String.format("Failed to migrate vm: %s" , vm.getUuid(), e);
+            s_logger.debug(msg);
+            throw new CloudRuntimeException(msg);
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: what about re throwing the exception? So, we do not loose the initial stack trace.
I mean, you can send the original exception in the new `CloudRuntimeException`

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 16]
```diff
@@ -1988,92 +1977,249 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
...
+        } finally {
+            try {
+                stateTransitTo(vm, Event.AgentReportStopped, null);
+            } catch (final NoTransitionException e) {
+                String anotherMsg = String.format("Failed to change vm state of %s", vmUuid, e);
+                s_logger.debug(anotherMsg);
+                throw new CloudRuntimeException(anotherMsg);
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: The same here

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 15 above)
Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Good point, changed to "throw new CloudRuntimeException(msg, e);" and removed the exception from the debug message.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 16 above)
Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: changed here as well

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java [hunk 17]
```diff
@@ -25,11 +25,28 @@
 import com.cloud.agent.api.to.VirtualMachineTO;
 import com.cloud.host.Host;
 
+/**
+ * interface to query how to move data around and to commision the moving
+ */
 public interface DataMotionStrategy {
+    /**
+     * reports whether this instance can do a move from source to destination
```
[the remark this answers is not in the document]
Comment (dhlaluku) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: Done

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java (hunk 5 above)
Comment (dhlaluku) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/DataMotionStrategy.java]: done

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 13 above)
Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Done

File: api/src/main/java/com/cloud/hypervisor/HypervisorGuru.java [hunk 18]
```diff
@@ -84,4 +85,13 @@
     List<Command> finalizeExpungeVolumes(VirtualMachine vm);
 
     Map<String, String> getClusterSettings(long vmId);
+
+    /**
+     * will generate commands to migrate a vm to a pool. For now this will ony work for stopped VMs on Vmware.
```
[the remark this answers is not in the document]
Comment (dhlaluku) [reply api/src/main/java/com/cloud/hypervisor/HypervisorGuru.java]: Corrected

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java [hunk 19]
```diff
@@ -119,13 +119,7 @@ public String getEventType() {
...
     @Override
     public String getEventDescription() {
-        if (getHostId() != null) {
-            return "Attempting to migrate VM Id: " + this._uuidMgr.getUuid(VirtualMachine.class, getVirtualMachineId()) + " to host Id: " + this._uuidMgr.getUuid(Host.class, getHostId());
-        } else if (getStoragePoolId() != null) {
-            return "Attempting to migrate VM Id: " + this._uuidMgr.getUuid(VirtualMachine.class, getVirtualMachineId()) + " to storage pool Id: " + this._uuidMgr.getUuid(StoragePool.class, getStoragePoolId());
-        } else {
-            return "Attempting to migrate VM Id: " + this._uuidMgr.getUuid(VirtualMachine.class, getVirtualMachineId());
-        }
+        return "Attempting to migrate VM Id: " + getVirtualMachineId() + " to host Id: " + getHostId();
```
[the remark this answers is not in the document]
Comment (nvazquez) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: Hi @mike-tutkowski, actually it was also possible to perform storage migration by setting the storagePoolId parameter

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 14 above)
Comment (nvazquez) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Agreed. How about this instead?

````
Commands commandsContainer = new Commands(Command.OnError.Stop);
commandsContainer.addCommands(commandsToSend);
````

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (mike-tutkowski) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: Thanks, @nvazquez! Do you know why we then removed the if/else branching in favor of just one string? Will getHostId() always be non null?

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-65

**Comment by dhlaluku** [comment]
@borisstoyanov thanks for investigating this, I will have a look.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 20]
```diff
@@ -4720,8 +4721,11 @@ protected VirtualMachine retrieve() {
         final User user = context.getCallingUser();
         final Account account = context.getCallingAccount();
 
-        for (StoragePool destPool : dest.getStorageForDisks().values()) {
-            checkConcurrentJobsPerDatastoreThreshhold(destPool);
+        Collection<StoragePool> storagePools = dest.getStorageForDisks().values();
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: so how about when dest or dest.getStorageForDisks() is null, this would still give a npe, no?

Comment (dhlaluku) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: That's true, the nullity check has to be for the Map returned by 'dest.getStorageForDisks()'.
Refactored to use this approach;

        Map<Volume, StoragePool> volumeStorageMap = dest.getStorageForDisks();
        if (volumeStorageMap != null) {
            for (Volume vol : volumeStorageMap.keySet()) {
                checkConcurrentJobsPerDatastoreThreshhold(volumeStorageMap.get(vol));
            }
        }

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65

**Comment by dhlaluku** [comment]
@blueorangutan test matrix

**Comment by GabrielBrascher** [comment]
It would be nice to have this one on 4.12.0. Any update on the tests @borisstoyanov @DaanHoogland?

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (mike-tutkowski) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: Thanks, @nvazquez! Do you know why we then removed the if/else branching in favor of just one string? Will getHostId() always be non null?

[the remark this answers is not in the document]
Comment (dhlaluku) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVMCmd.java]: @mike-tutkowski I have re-analyzed this and getHostID() can be null considering that a VM can either be migrated to a host or storage pool, both of which are optional parameters. I will revert back this change to cater for the migrate VM to either host or storage pool. /CC @DaanHoogland

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65

**Comment by dhlaluku** [comment]
@rafaelweingartner do we have any outstanding items for this? /CC @GabrielBrascher

**Comment by dhlaluku** [comment]
`nosetests --with-xunit --xunit-file=results.xml --with-marvin --marvin-config=setup/dev/advanced.cfg -s -a tags=advanced,require_hardware="false" --zone=Sandbox-Simulator --hypervisor=simulator test/integration/smoke/test_primary_storage.py

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp/MarvinLogs/Jan_23_2019_09_14_40_6K9K05. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
=== TestName: test_01_add_primary_storage_disabled_host | Status : SUCCESS ===

=== TestName: test_01_primary_storage_nfs | Status : SUCCESS ===

=== TestName: test_01_deploy_vms_storage_tags | Status : SUCCESS ===

=== TestName: test_02_edit_primary_storage_tags | Status : SUCCESS ===

=== TestName: test_03_migration_options_storage_tags | Status : SUCCESS ===

===final results are now copied to: /tmp/MarvinLogs/test_primary_storage_NFVHD7===
`
@rafaelweingartner not sure why Travis CI is failing but tests are passing from my local simulator environment

**Comment by rafaelweingartner** [comment]
The error changed now. did you rebase or something?
```
Traceback (most recent call last):
  File "/home/travis/.local/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/home/travis/.local/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/home/travis/.local/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/home/travis/.local/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/home/travis/build/apache/cloudstack/test/integration/smoke/test_deploy_vm_root_resize.py", line 55, in setUpClass
    cls.services["mode"] = cls.zone.networktype
AttributeError: 'str' object has no attribute 'networktype'
```

**Comment by dhlaluku** [comment]
@rafaelweingartner I did a rebase against the latest master and force-pushed. The Travis CI errors are not related to this PR and they keep on changing every time I do a force-push. I have tested them locally against simulator based env with Ubuntu 18.04. 

It is probably worth noting that they might be related to the environments with which they run, I notice that they run against Debian trusty ~ Ubuntu 14.04. We should probably upgrade to at least Ubuntu 16.04 LTS

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by dhlaluku** [comment]
@blueorangutan test centos7 vmware-65
